### PR TITLE
Add a Toggle Between Commands Binding

### PR DESCRIPTION
**Pull request by litehed**
Added a method to toggle between two different commands Also can be used with runnables

# Pull Requests

Please note that we accept pull requests from anyone, but that does not mean it will be merged.

## What kind of change does this PR introduce?
* Feature - added a method in Trigger/Button which allows a person to toggle between two commands
I created three new methods with the name toggleWhenActive(Trigger)/toggleWhenPressed(Button). All they simply do is they take in two commands or runnables and utilize the same code as the single command toggleWhenxxxx except they check which command was scheduled last and based upon that, it schedules the other command. Still needs to be tested by the way, but it looks like it should work fine.
## Did this PR introduce a breaking change?
* No

__Please make sure your PR satisfies the requirements of [the contributing page](CONTRIBUTING.md)__

### CODE REVIEW

File: core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java [hunk 1]
```diff
@@ -283,6 +283,64 @@ public Trigger toggleWhenActive(final Command command) {
...
+     * Toggles between two commands when the trigger becomes active (commadOne then commandTwo
+     * then commandOne). These commands are set to be interruptible.
+     *
+     * @param commandOne the command to start
+     * @param commandTwo the command to be activated after
+     * @return this trigger, so calls can be chained
+     */
+    public Trigger toggleWhenActive(final Command commandOne, final Command commandTwo){
+        toggleWhenActive(commandOne, commandTwo, true);
+        return this;
```
Comment (JIceberg) [inline core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java:328]: This should just be `return toggleWhenActive(commandOne, commandTwo, true)`. This should also be the case for the rest of the methods that utilize another method.

File: core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java [hunk 2]
```diff
@@ -283,6 +283,64 @@ public Trigger toggleWhenActive(final Command command) {
...
+                    } else if(commandTwo.isScheduled()){
+                        commandTwo.cancel();
+                        commandOne.schedule(interruptible);
+                    } else {
+                        commandOne.schedule(interruptible);
+                    }
+                }
+                m_pressedLast = pressed;
+            }
+        });
```
Comment (JIceberg) [inline core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java:314]: There's a bit of difference in whitespace here that's different from the other methods. Refer to them for the proper spacing.

File: core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java [hunk 3]
```diff
@@ -283,6 +283,64 @@ public Trigger toggleWhenActive(final Command command) {
...
+                if (!m_pressedLast && pressed) {
+                    if (commandOne.isScheduled()) {
+                        commandOne.cancel();
+                        commandTwo.schedule(interruptible);
+                    } else if(commandTwo.isScheduled()){
+                        commandTwo.cancel();
+                        commandOne.schedule(interruptible);
+                    } else {
+                        commandOne.schedule(interruptible);
+                    }
```
Comment (JIceberg) [inline core/src/main/java/com/arcrobotics/ftclib/command/button/Trigger.java:310]: This wouldn't work if the command is cancelled after one iteration (e.g. an `InstantCommand`) or is cancelled before the next active state. I suggest using a boolean here that contains the state called `m_firstCommandActive` outside of the `run()` method. It should be instantiated to `false`. You can then limit this to just a simple if-else. Make sure your spacing is consistent. Do Ctrl+Alt+L to auto-format.

Below is how this should be done:
```java
if (m_firstCommandActive) {
  if (commandOne.isScheduled()) {
    commandOne.cancel();
  }
  commandTwo.schedule(interruptible);
} else {
  if (commandTwo.isScheduled()) {
    commandTwo.cancel();
  }
  commandOne.schedule(interruptible);
}
m_firstCommandActive = !m_firstCommandActive;

**Review by JIceberg — CHANGES_REQUESTED** [review]

**Comment by JIceberg** [comment]
I also need you to pull changes before making another commit. We just merged two pull requests (none that change the content of your PR, but you should still pull before pushing).

**Comment by JIceberg** [comment]
@litehed You didn't check your stashed changes.
image

Please clear those and re-push.

**Comment by litehed** [comment]
> @litehed You didn't check your stashed changes.
> image
> 
> Please clear those and re-push.

Yeah I noticed that instantly Im doing that

**Comment by JIceberg** [comment]
Next step is writing a quick unit test. Based on the method, it should toggle between two commands. This can be tested pretty easily. Go to `CommandSchedulerTests` and add these tests (remove the `x = 3;`, `enable()`, and `reset()` calls from the beginning/end of all tests):
```java
@BeforeEach
public void setup() {
  x = 3;
  Robot.enable();
}

@AfterEach
public void teardown() {
  CommandScheduler.getInstance().reset();
}
```
Then add this test for your trigger binding:
```java
/**
 * Tests the toggle binding that switches between two commands.
 * Runs two iterations of the toggle binding.
 */
@Test
public void testToggleBetweenCommands() {
  Trigger toggler = new Trigger(this::getValue).toggleWhenActive(
    new InstantCommand(() -> x = 5),
    new InstantCommand(() -> x = 3)
  );
  assertEquals(3, x);
  CommandScheduler.getInstance().run();
  assertEquals(3, x, "The value of x should have updated");

  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(5, x);
  CommandScheduler.getInstance().run();
  assertEquals(5, x, "The value of x should not have updated");

  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(5, x, "The value of x should not have updated");
  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(3, x, "The value of x should have updated");

  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(3, x, "The value of should not have updated");
  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(5, x, "The value of x should have updated");

  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(5, x, "The value of should not have updated");
  updateValue();
  CommandScheduler.getInstance().run();
  assertEquals(3, x, "The value of x should have updated"); 
}

**Comment by JIceberg** [comment]
Slight issue I have with the `CommandSchedulerTests` update.. the spacing. Can you do ctrl+alt+L on that file and make another commit?

**Review by JIceberg — APPROVED** [review]

**Review by JarnaChao09 — APPROVED** [review]

**Pull request merged**